src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,744 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NLog.Internal
+{
+    /// <summary>
+    /// Dictionary that combines the standard <see cref="LogEventInfo.Properties" /> with the
+    /// MessageTemplate-properties extracted from the <see cref="LogEventInfo.Message" />.
+    /// 
+    /// The <see cref="MessageProperties" /> are returned as the first items
+    /// in the collection, and in positional order.
+    /// </summary>
+    internal sealed class PropertiesDictionary : IDictionary<object, object>, IDictionary
+    {
+        struct PropertyValue
+        {
+            public readonly object Value;
+            public readonly bool MessageProperty;
+
+            public PropertyValue(object value, bool messageProperty)
+            {
+                Value = value;
+                MessageProperty = messageProperty;
+            }
+        }
+
+        private Dictionary<object, PropertyValue> _eventProperties;
+        private IList<MessageTemplateParameter> _messageProperties;
+        private DictionaryCollection _keyCollection;
+        private DictionaryCollection _valueCollection;
+
+        private bool IsEmpty { get { return (_eventProperties == null || _eventProperties.Count == 0) && (_messageProperties == null || _messageProperties.Count == 0); } }
+
+        private Dictionary<object, PropertyValue> EventProperties
+        {
+            get
+            {
+                if (_eventProperties == null)
+                {
+                    if (_messageProperties != null && _messageProperties.Count > 0)
+                    {
+                        _eventProperties = new Dictionary<object, PropertyValue>(_messageProperties.Count);
+                        if (!InsertMessagePropertiesDictionary(_messageProperties, _eventProperties))
+                        {
+                            _messageProperties = CreateUniqueMessageProperties(_messageProperties, _eventProperties);
+                        }
+                    }
+                    else
+                    {
+                        _eventProperties = new Dictionary<object, PropertyValue>();
+                    }
+                }
+                return _eventProperties;
+            }
+        }
+
+        internal IList<MessageTemplateParameter> MessageProperties
+        {
+            get
+            {
+                return _messageProperties;
+            }
+        }
+
+        internal void SetMessageTemplateParameterList(IMessageTemplateParameterList value)
+        {
+            ResetMessageProperties();
+
+            if (value != null && value.Count > 0)
+            {
+                if (_eventProperties == null && value.Count <= 10)
+                {
+                    bool uniqueMessageProperties = true;
+                    for (int i = 0; i < value.Count - 1; ++i)
```

Comment:
What's uniqueMessageProperties  here? What are we trying to do here?

Reply:
Ensure the MessageTemplate-parameters can be used as Dictionary-values. Ex. "Male user: {userid} loves female user: {userid}".